src/service/logs/ingest.rs

```diff
@@ -274,19 +266,26 @@ pub async fn ingest(
             }
         };
 
-        let function_no = stream_before_functions_map
-            .get(&main_stream_key)
-            .map(|v| v.len())
-            .unwrap_or_default()
-            + stream_after_functions_map
-                .get(&key)
-                .map(|v| v.len())
-                .unwrap_or_default();
-        let (ts_data, fn_num) = json_data_by_stream
-            .entry(routed_stream_name.clone())
-            .or_insert((Vec::new(), None));
-        ts_data.push((timestamp, local_val));
-        *fn_num = need_usage_report.then_some(function_no);
```

Comment:
This get/update on the `json_data_by_stream`, which stores both data assigned to each stream and function usage statistics for each stream, has now been split into two places.

This old code performed the get/update on `json_data_by_stream` and update the assigned data and function usage statistics at the exact same time.

The new code updates the function usage statistics first, immediately after functions are applied to data. Then, later, the new code updates the assigned data for each destination stream that the data will be routed to. (These destination streams may or may not include the original stream).

Since assigned data and function usage statistics are now updated independently of one another, it may no longer be desirable to store both assigned data and function usage statistics as values in the same `HashMap`. Rather, it might make the code more readable to introduce a new `fn_usage_by_stream` `HashMap` that tracks function usage statistics separately from assigned data. This could be a refactoring for a followup PR.